### PR TITLE
Add support for alternate certificate download

### PR DESCRIPTION
**Pull request by unixcharles**
This PR include

- A class to identify and match the chain
- Support for a `preferred_chain` optional karg on `Client#certificate`
- `preferred_chain` forward in `Order#certificate`

The only thing I'm not 100% is `preferred_chain` as a name. Its following certbot naming but it might confused people into thinking that it will try to find the prefer chain and fallback to something else if not found.

This implementation raise if the chain does not match. I think it would be up to the library user to implement fallback mechanism.

i.e.;

```
order.certificate(preferred_chain: 'IdenTrust') || order.certificate
```

### CODE REVIEW

**Comment by unixcharles** [comment]
cc @haeky

**Comment by unixcharles** [comment]
@sawanoboly @ryansouza went the direction of `#download(url, preferred_chain:`.

Do you have a use case for fetching the retrieving the alt and not just the matching one?

**Comment by sawanoboly** [comment]
@unixcharles Thanks for the implementation 👍 

> Do you have a use case for fetching the retrieving the alt and not just the matching one?

I can't envision such a use case at the moment. If it does occur, we'll think about it again when it happens.

**Comment by sawanoboly** [comment]
I checked the operation with the endpoint `https://acme-v02.api.letsencrypt.org/directory`.

It is working, thank you!


default => `DST Root CA X3`  (until Sep 29)

```
> order.certificate
=> "-----BEGIN CERTIFICATE-----\nMIIFYDCCBEigAwIBAgISBOVeu3w3Oj8...

...

> OpenSSL::X509::Certificate.new(Intermediate_certificate).issuer
=> #<OpenSSL::X509::Name CN=DST Root CA X3,O=Digital Signature Trust Co.>
```

`preferred_chain: 'DST Root CA X3'` => `DST Root CA X3`

```
> order.certificate(preferred_chain: 'DST Root CA X3')
=> "-----BEGIN CERTIFICATE-----\nMIIFYDCCBEigAwIBAgISBOVeu3w3Oj8PNak...

...

> OpenSSL::X509::Certificate.new(Intermediate_certificate2).issuer
=> #<OpenSSL::X509::Name CN=DST Root CA X3,O=Digital Signature Trust Co.>
```

`preferred_chain: 'ISRG Root X1''` => `ISRG Root X1'`

```
> order.certificate(preferred_chain: 'ISRG Root X1')
=> "-----BEGIN CERTIFICATE-----\nMIIFYDCCBEigAwIBAgIS...

...

> OpenSSL::X509::Certificate.new(Intermediate_certificate3).issuer
=> #<OpenSSL::X509::Name CN=DST Root CA X3,O=Digital Signature Trust Co.>
```

**Comment by sawanoboly** [comment]
I checked the operation with the endpoint `https://acme-v02.api.letsencrypt.org/directory`.

I have one question.

default => `DST Root CA X3`  (until Sep 29)

```
> order.certificate
=> "-----BEGIN CERTIFICATE-----\nMIIFYDCCBEigAwIBAgISBOVeu3w3Oj8...

...

> OpenSSL::X509::Certificate.new(Intermediate_certificate).issuer
=> #<OpenSSL::X509::Name CN=DST Root CA X3,O=Digital Signature Trust Co.>
```

`preferred_chain: 'DST Root CA X3'` => `DST Root CA X3`

```
> order.certificate(preferred_chain: 'DST Root CA X3')
=> "-----BEGIN CERTIFICATE-----\nMIIFYDCCBEigAwIBAgISBOVeu3w3Oj8PNak...

...

> OpenSSL::X509::Certificate.new(Intermediate_certificate2).issuer
=> #<OpenSSL::X509::Name CN=DST Root CA X3,O=Digital Signature Trust Co.>
```

`preferred_chain: 'ISRG Root X1''` => `DST Root CA X3` ?

```
> order.certificate(preferred_chain: 'ISRG Root X1')
=> "-----BEGIN CERTIFICATE-----\nMIIFYDCCBEigAwIBAgIS...

...

> OpenSSL::X509::Certificate.new(Intermediate_certificate3).issuer
=> #<OpenSSL::X509::Name CN=DST Root CA X3,O=Digital Signature Trust Co.>
```

Why is this a match?

```
> chain_identifier= Acme::Client::ChainIdentifier.new(order.certificate(preferred_chain: 'ISRG Root X1'))
> chain_identifier.send(:issuers)
=> ["/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3", "/O=Digital Signature Trust Co./CN=DST Root CA X3"]
```

**Comment by unixcharles** [comment]
Can you gist the full certificate chain in PEM format so I can give it a try?

**Comment by unixcharles** [comment]
Testing with `https://shopify.com` certificate chain:

```
-----BEGIN CERTIFICATE-----
MIIFTTCCBDWgAwIBAgISA2AnA6wRtECFIeRayASvDeMXMA0GCSqGSIb3DQEBCwUA
MEoxCzAJBgNVBAYTAlVTMRYwFAYDVQQKEw1MZXQncyBFbmNyeXB0MSMwIQYDVQQD
ExpMZXQncyBFbmNyeXB0IEF1dGhvcml0eSBYMzAeFw0yMDA4MjUxNjI3MDBaFw0y
MDExMjMxNjI3MDBaMBYxFDASBgNVBAMTC3Nob3BpZnkuY29tMIIBIjANBgkqhkiG
9w0BAQEFAAOCAQ8AMIIBCgKCAQEAsNlaIAVy/SvQZGSvKOjvDwRQ0c6fRvEs6Pht
dYSz1N9FzvtUEFp2QW+42rPOP6ZW4LpWiT8UzIEl+XePkSYAeymlrC7b5ZdogudX
4WEAptJ9drYfvRXcV+z5HQi3xTBSzLErXOwiqNwYcl3jRffLNzfyYqcvkSlFHhsp
YlO7q/Eviqjw1wE+v5X8qHlr6HXmcNxAreg5DHXHvkEqxy+onxa1IyRTIbs2/a7y
hM7c19XJGk7frJkSnRr7hk/XaJg5O0ToGWS9PyTQ1Jq/Y7yb8gaX/d525aUeUfOA
yGh8fl5k31ak/NAdD063KnPSBUcrR8818NS2P70FRwSmVoZ7cwIDAQABo4ICXzCC
AlswDgYDVR0PAQH/BAQDAgWgMB0GA1UdJQQWMBQGCCsGAQUFBwMBBggrBgEFBQcD
AjAMBgNVHRMBAf8EAjAAMB0GA1UdDgQWBBSEt5q9jbh0vJZgfWrS4zMeXRL0UzAf
BgNVHSMEGDAWgBSoSmpjBH3duubRObemRWXv86jsoTBvBggrBgEFBQcBAQRjMGEw
LgYIKwYBBQUHMAGGImh0dHA6Ly9vY3NwLmludC14My5sZXRzZW5jcnlwdC5vcmcw
LwYIKwYBBQUHMAKGI2h0dHA6Ly9jZXJ0LmludC14My5sZXRzZW5jcnlwdC5vcmcv
MBYGA1UdEQQPMA2CC3Nob3BpZnkuY29tMEwGA1UdIARFMEMwCAYGZ4EMAQIBMDcG
CysGAQQBgt8TAQEBMCgwJgYIKwYBBQUHAgEWGmh0dHA6Ly9jcHMubGV0c2VuY3J5
cHQub3JnMIIBAwYKKwYBBAHWeQIEAgSB9ASB8QDvAHUA5xLysDd+GmL7jskMYYTx
6ns3y1YdESZb8+DzS/JBVG4AAAF0JqimhAAABAMARjBEAiAjR40Y+i24/spHhSyI
zHxiGmbBJKWl2uHQPWyDv1atFwIgeneyrj5i7ikXCrurYiT3yLtGHELt9Hm6HnPP
KKvz0aMAdgAHt1wb5X1o//Gwxh0jFce65ld8V5S3au68YToaadOiHAAAAXQmqKap
AAAEAwBHMEUCIDcPNW1zbVCvmmPvpk0xxpYqolLDrVn5fpAk58L6lOiUAiEAkglQ
kr0C2AHm/ZxXVchJPottpm0+3se2ZJw6h8yanQowDQYJKoZIhvcNAQELBQADggEB
ABmfVYAnJmeJtRt/bMH6mxkS9F8ph7As9pBKpruwoUOe09IXyez2lzIvpYlBHiwJ
i1AWCEXRE+JXg7rgPYDqemiB7LFZTdy6zTEjAby5JpkZUihIDtmDAuForj8+tYiJ
C9ipbg1547Bd8DQcayU/w0bqJAMrtE7AqpvnJOGDys2EUWRbjOECAZkgVReFRzmZ
lOPAAGfXS3l7BvcnKsNr9wpbwy+zZucCgwSxAuZME/98nBFG6Y0/O4+ZOJLPjcxv
VWKA8W116wnrMKwKcLe5hLlLjbd5sdI19C16jakPZsUN8nIbvUjq/lrqEZlSqhKF
dCZOWz+Xeegy11OI444VNN4=
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
MIIEkjCCA3qgAwIBAgIQCgFBQgAAAVOFc2oLheynCDANBgkqhkiG9w0BAQsFADA/
MSQwIgYDVQQKExtEaWdpdGFsIFNpZ25hdHVyZSBUcnVzdCBDby4xFzAVBgNVBAMT
DkRTVCBSb290IENBIFgzMB4XDTE2MDMxNzE2NDA0NloXDTIxMDMxNzE2NDA0Nlow
SjELMAkGA1UEBhMCVVMxFjAUBgNVBAoTDUxldCdzIEVuY3J5cHQxIzAhBgNVBAMT
GkxldCdzIEVuY3J5cHQgQXV0aG9yaXR5IFgzMIIBIjANBgkqhkiG9w0BAQEFAAOC
AQ8AMIIBCgKCAQEAnNMM8FrlLke3cl03g7NoYzDq1zUmGSXhvb418XCSL7e4S0EF
q6meNQhY7LEqxGiHC6PjdeTm86dicbp5gWAf15Gan/PQeGdxyGkOlZHP/uaZ6WA8
SMx+yk13EiSdRxta67nsHjcAHJyse6cF6s5K671B5TaYucv9bTyWaN8jKkKQDIZ0
Z8h/pZq4UmEUEz9l6YKHy9v6Dlb2honzhT+Xhq+w3Brvaw2VFn3EK6BlspkENnWA
a6xK8xuQSXgvopZPKiAlKQTGdMDQMc2PMTiVFrqoM7hD8bEfwzB/onkxEz0tNvjj
/PIzark5McWvxI0NHWQWM6r6hCm21AvA2H3DkwIDAQABo4IBfTCCAXkwEgYDVR0T
AQH/BAgwBgEB/wIBADAOBgNVHQ8BAf8EBAMCAYYwfwYIKwYBBQUHAQEEczBxMDIG
CCsGAQUFBzABhiZodHRwOi8vaXNyZy50cnVzdGlkLm9jc3AuaWRlbnRydXN0LmNv
bTA7BggrBgEFBQcwAoYvaHR0cDovL2FwcHMuaWRlbnRydXN0LmNvbS9yb290cy9k
c3Ryb290Y2F4My5wN2MwHwYDVR0jBBgwFoAUxKexpHsscfrb4UuQdf/EFWCFiRAw
VAYDVR0gBE0wSzAIBgZngQwBAgEwPwYLKwYBBAGC3xMBAQEwMDAuBggrBgEFBQcC
ARYiaHR0cDovL2Nwcy5yb290LXgxLmxldHNlbmNyeXB0Lm9yZzA8BgNVHR8ENTAz
MDGgL6AthitodHRwOi8vY3JsLmlkZW50cnVzdC5jb20vRFNUUk9PVENBWDNDUkwu
Y3JsMB0GA1UdDgQWBBSoSmpjBH3duubRObemRWXv86jsoTANBgkqhkiG9w0BAQsF
AAOCAQEA3TPXEfNjWDjdGBX7CVW+dla5cEilaUcne8IkCJLxWh9KEik3JHRRHGJo
uM2VcGfl96S8TihRzZvoroed6ti6WqEBmtzw3Wodatg+VyOeph4EYpr/1wXKtx8/
wApIvJSwtmVi4MFU5aMqrSDE6ea73Mj2tcMyo5jMd6jmeWUHK8so/joWUoHOUgwu
X4Po1QYz+3dszkDqMp4fklxBwXRsW10KXzPMTZ+sOPAveyxindmjkW8lGy+QsRlG
PfZ+G6Z6h7mjem0Y+iWlkYcV4PIWL1iwBi8saCbGS5jN2p8M+X+Q7UNKEkROb3N6
KOqkqm57TH2H3eDJAkSnh6/DNFu0Qg==
-----END CERTIFICATE-----
```

Loading this in `Acme::Client::ChainIdentifier`

```
> chain_identifier = Acme::Client::ChainIdentifier.new(pem)
=> "..."
> chain_identifier.send :issuers
=> ["/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3", "/O=Digital Signature Trust Co./CN=DST Root CA X3"]
> chain_identifier.match_name?("ISRG Root X1")
=> false
```

So unless there is a logic problem with `certificate` method itself it should work.

**Comment by sawanoboly** [comment]
Presumably under the current state of affairs, `(preferred_chain: 'ISRG Root X1'` should raise a PreferredChainNotMatched exception, but in fact it doesn't.

```
> order.certificate(preferred_chain: 'FOOBAR Root X1')
Acme::Client::Error::PreferredChainNotMatched: Could not find any matching chain for `FOOBAR Root X1`
```

```
> fullchain = order.certificate(preferred_chain: 'ISRG Root X1')
> puts fullchain

-----BEGIN CERTIFICATE-----
MIIFXjCCBEagAwIBAgISBE7WroXyu+LxGeCd4gK9q58JMA0GCSqGSIb3DQEBCwUA
MEoxCzAJBgNVBAYTAlVTMRYwFAYDVQQKEw1MZXQncyBFbmNyeXB0MSMwIQYDVQQD
ExpMZXQncyBFbmNyeXB0IEF1dGhvcml0eSBYMzAeFw0yMDA5MTQwMDQ3MzRaFw0y
MDEyMTMwMDQ3MzRaMB8xHTAbBgNVBAMTFGFhYWExOC5vcHNyb2NraW4uY29tMIIB
IjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAuUHtM+PRajeq7/NDWbzeBSLR
Ne0YcU0DD0WChYL2eX+wuWFaBf7LWv3LD/KXr8ZvgNM97eq6j6K+BWp/aVHt299x
Ggt/XF+rpjY6gl7iL6PRC8s6FZf2e7HLD4h110VaTVXpSBYPTrlcpxU4tho6JWgG
EMBZFqw3sfOFbbCKz/GRECbTeq7PAWqOHMKnAtxY+ZGdH62l4izoka1gT1+6RVPb
RsizF10o1AOwFUj47hU1yO1vqESmbnmPF8We90S1sOwg5ZSN+4FZ1CcPrCu2XVqR
Lc0oamQ8QRBarLCB+nsDL7bRltomg2PZrbOcx2cwkT31tIOE1580UKL5ELxf/wID
AQABo4ICZzCCAmMwDgYDVR0PAQH/BAQDAgWgMB0GA1UdJQQWMBQGCCsGAQUFBwMB
BggrBgEFBQcDAjAMBgNVHRMBAf8EAjAAMB0GA1UdDgQWBBSO050W2qwO3H+ikyx/
dsylJcmiRTAfBgNVHSMEGDAWgBSoSmpjBH3duubRObemRWXv86jsoTBvBggrBgEF
BQcBAQRjMGEwLgYIKwYBBQUHMAGGImh0dHA6Ly9vY3NwLmludC14My5sZXRzZW5j
cnlwdC5vcmcwLwYIKwYBBQUHMAKGI2h0dHA6Ly9jZXJ0LmludC14My5sZXRzZW5j
cnlwdC5vcmcvMB8GA1UdEQQYMBaCFGFhYWExOC5vcHNyb2NraW4uY29tMEwGA1Ud
IARFMEMwCAYGZ4EMAQIBMDcGCysGAQQBgt8TAQEBMCgwJgYIKwYBBQUHAgEWGmh0
dHA6Ly9jcHMubGV0c2VuY3J5cHQub3JnMIIBAgYKKwYBBAHWeQIEAgSB8wSB8ADu
AHUAsh4FzIuizYogTodm+Su5iiUgZ2va+nDnsklTLe+LkF4AAAF0ikvDGgAABAMA
RjBEAiAWqqfjz8sOiPY3G3KFTYPFHz1gdHm6CiQIZ2A0vg6DlgIgTWgexcFbK0yk
IWyHblGEPYPugzwxcUHQ69iAgE/R80IAdQBep3P531bA57U2SH3QSeAyepGaDISh
EhKEGHWWgXFFWAAAAXSKS8NPAAAEAwBGMEQCIB3sDPJuWMN8uZ8nGfyGCOoEitC7
7QKF4Hik75B4AkZVAiBf1ZPIutuG8pnfcQGUkA2vAE/je6x9K6aWc8ptrtpjnjAN
BgkqhkiG9w0BAQsFAAOCAQEAeLbPa/ThKlRDmN90pAKv58FkAA05mVNi7nIGVlql
rn62gBpm7i6eMq+2bHJ2ExiNCqCuEATLAHL4HcYYGu8FFrkyNqzzy3RK/mWfe1P+
Y7AtXg3OH6+6XSqmHltu1WJ2fhlhHFqJp0dXQZYoU5nB5k7Oa87aYGleAKu+cwH/
ec4FdYecwn5uEbJxPY7p4obwJq6kPmj9mYA4FB5Wkyea91kWXo4IyyFtHYWjm/eh
aGDv1HTomV5pisdfrvwNHa7IoBKEczYyISxNKPx+qX3oAmDWmbxpECNAu5TA5WzJ
0K5EvcXvvFIc991S8YcGCJw3XMakHVlv/c4b6k/NBG3pkA==
-----END CERTIFICATE-----

-----BEGIN CERTIFICATE-----
MIIEkjCCA3qgAwIBAgIQCgFBQgAAAVOFc2oLheynCDANBgkqhkiG9w0BAQsFADA/
MSQwIgYDVQQKExtEaWdpdGFsIFNpZ25hdHVyZSBUcnVzdCBDby4xFzAVBgNVBAMT
DkRTVCBSb290IENBIFgzMB4XDTE2MDMxNzE2NDA0NloXDTIxMDMxNzE2NDA0Nlow
SjELMAkGA1UEBhMCVVMxFjAUBgNVBAoTDUxldCdzIEVuY3J5cHQxIzAhBgNVBAMT
GkxldCdzIEVuY3J5cHQgQXV0aG9yaXR5IFgzMIIBIjANBgkqhkiG9w0BAQEFAAOC
AQ8AMIIBCgKCAQEAnNMM8FrlLke3cl03g7NoYzDq1zUmGSXhvb418XCSL7e4S0EF
q6meNQhY7LEqxGiHC6PjdeTm86dicbp5gWAf15Gan/PQeGdxyGkOlZHP/uaZ6WA8
SMx+yk13EiSdRxta67nsHjcAHJyse6cF6s5K671B5TaYucv9bTyWaN8jKkKQDIZ0
Z8h/pZq4UmEUEz9l6YKHy9v6Dlb2honzhT+Xhq+w3Brvaw2VFn3EK6BlspkENnWA
a6xK8xuQSXgvopZPKiAlKQTGdMDQMc2PMTiVFrqoM7hD8bEfwzB/onkxEz0tNvjj
/PIzark5McWvxI0NHWQWM6r6hCm21AvA2H3DkwIDAQABo4IBfTCCAXkwEgYDVR0T
AQH/BAgwBgEB/wIBADAOBgNVHQ8BAf8EBAMCAYYwfwYIKwYBBQUHAQEEczBxMDIG
CCsGAQUFBzABhiZodHRwOi8vaXNyZy50cnVzdGlkLm9jc3AuaWRlbnRydXN0LmNv
bTA7BggrBgEFBQcwAoYvaHR0cDovL2FwcHMuaWRlbnRydXN0LmNvbS9yb290cy9k
c3Ryb290Y2F4My5wN2MwHwYDVR0jBBgwFoAUxKexpHsscfrb4UuQdf/EFWCFiRAw
VAYDVR0gBE0wSzAIBgZngQwBAgEwPwYLKwYBBAGC3xMBAQEwMDAuBggrBgEFBQcC
ARYiaHR0cDovL2Nwcy5yb290LXgxLmxldHNlbmNyeXB0Lm9yZzA8BgNVHR8ENTAz
MDGgL6AthitodHRwOi8vY3JsLmlkZW50cnVzdC5jb20vRFNUUk9PVENBWDNDUkwu
Y3JsMB0GA1UdDgQWBBSoSmpjBH3duubRObemRWXv86jsoTANBgkqhkiG9w0BAQsF
AAOCAQEA3TPXEfNjWDjdGBX7CVW+dla5cEilaUcne8IkCJLxWh9KEik3JHRRHGJo
uM2VcGfl96S8TihRzZvoroed6ti6WqEBmtzw3Wodatg+VyOeph4EYpr/1wXKtx8/
wApIvJSwtmVi4MFU5aMqrSDE6ea73Mj2tcMyo5jMd6jmeWUHK8so/joWUoHOUgwu
X4Po1QYz+3dszkDqMp4fklxBwXRsW10KXzPMTZ+sOPAveyxindmjkW8lGy+QsRlG
PfZ+G6Z6h7mjem0Y+iWlkYcV4PIWL1iwBi8saCbGS5jN2p8M+X+Q7UNKEkROb3N6
KOqkqm57TH2H3eDJAkSnh6/DNFu0Qg==
-----END CERTIFICATE-----
```

The common name of the certificate obtained is `DST Root CA X3`.

```
> Acme::Client::ChainIdentifier.new(fullchain).send :issuers
=> ["/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3", "/O=Digital Signature Trust Co./CN=DST Root CA X3"]
```

I understand that the ChainIdentifier is not very strict. However, it does not seem to be a problem in and of itself.

```
> Acme::Client::ChainIdentifier.new(fullchain).match_name?("ISRG Root X1")
=> false
```

```
# case sensitive
> order.certificate(preferred_chain: 'ISRG ROOT X1')
Acme::Client::Error::PreferredChainNotMatched: Could not find any matching chain for `ISRG ROOT X1`
```

It almost seems to work. As I recall, the current CA certificate is a cross-route, so it's not a problem if we can get it. However, from the current implementation, it appears to be a strange behavior.

**Comment by unixcharles** [comment]
I'm returning the wrong `pem` 🤦

**Comment by unixcharles** [comment]
Should be fixed now

File: lib/acme/client.rb [hunk 1]
```diff
@@ -127,9 +128,28 @@ def finalize(url:, csr:)
     Acme::Client::Resources::Order.new(self, **arguments)
   end
 
-  def certificate(url:)
+  def certificate(url:, preferred_chain: nil)
     response = download(url, format: :pem)
-    response.body
+    pem = response.body
+    identifier_class = Acme::Client::ChainIdentifier
```
Comment (jhass) [inline lib/acme/client.rb:134]: Why the local over just using `ChainIdentifier` directly? We're in the `Acme::Client` namespace, so no need to prefix it. The local made me wonder if this is for some reason is configurable.

File: lib/acme/client/chain_identifier.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      certificate_lines = []
+      @pem_certificate_chain.each_line do |line|
+        certificate_lines << line
+        if line.include?('END CERTIFICATE')
+          certificates << certificate_lines.join
+          certificate_lines = []
+        end
+      end
+
+      certificates
```
Comment (jhass) [inline lib/acme/client/chain_identifier.rb:34]:
```suggestion
    def splitted_pem_certificates
      @pem_certificate_chain.each_line.slice_after(/END CERTIFICATE/)
```

**Comment by jsaubry** [comment]
It would be worth the effort adding a new entry in the documentation under `Downloading a certificate`. As you mentioned, I think the word `preferred` is a little bit misleading if it doesn't have a fallback, so it would be a good idea to mention that it's up to the user to implement the fallback.

File: lib/acme/client/chain_identifier.rb (hunk 2 above)
Comment (unixcharles) [reply lib/acme/client/chain_identifier.rb:34]: I like that

File: lib/acme/client.rb (hunk 1 above)
Comment (unixcharles) [reply lib/acme/client.rb:134]: You are right, no need to fully qualify here.

File: lib/acme/client/chain_identifier.rb (hunk 2 above)
Comment (jhass) [reply lib/acme/client/chain_identifier.rb]: Okay, I didn't expect Ruby 2.1 still be on the support list, given its official security maintenance phase is long over.

**Comment by gschorkopf** [comment]
@unixcharles inre:

> Do you have a use case for fetching the retrieving the alt and not just the matching one?

We're just looking for the matching one, so this solution should work!

**Review by jhass — APPROVED** [review]

**Comment by chrisUsick** [comment]
@unixcharles @jsaubry I have a PR to add documentation to this PR [here](https://github.com/unixcharles/acme-client/pull/192/files). Does this suffice?

**Comment by gschorkopf** [comment]
Looks great @unixcharles, thanks for your work on this!

File: lib/acme/client/chain_identifier.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    def issuers
+      x509_certificates.map(&:issuer).map(&:to_s)
+    end
+
+    def x509_certificates
+      @x509_certificates ||= splitted_pem_certificates.map { |pem| OpenSSL::X509::Certificate.new(pem) }
+    end
+
+    def splitted_pem_certificates
+      @pem_certificate_chain.each_line.slice_after(/END CERTIFICATE/).map(&:join)
```
Comment (gschorkopf) [inline lib/acme/client/chain_identifier.rb:24]: This is neat!

File: lib/acme/client.rb [hunk 4]
```diff
@@ -127,9 +128,27 @@ def finalize(url:, csr:)
...
+    pem = response.body
+
+    return pem if preferred_chain.nil?
+
+    return pem if ChainIdentifier.new(pem).match_name?(preferred_chain)
+    # TODO: Remove Array() after decode_link_headers fix
+    #
+    #   FaradayMiddleware#decode_link_headers a single entry
+    #   per rel= but nothing prevent it in the spec.
+    alternative_urls = Array(response.headers['link']['alternate'])
```
Comment (gschorkopf) [inline lib/acme/client.rb:142]: If you ask for a `preferred_chain`, and the CA does not respond with any `link`s, we need to protect against that here.

```suggestion
    alternative_urls = Array(response.headers.dig('link', 'alternate'))
```

Comment (gschorkopf) [reply lib/acme/client.rb:142]: I'm not sure if that's explicitly addressed somewhere here in your TODO or not https://github.com/unixcharles/acme-client/pull/191

**Review by gschorkopf — CHANGES_REQUESTED** [review]

File: lib/acme/client.rb (hunk 4 above)
Comment (unixcharles) [reply lib/acme/client.rb:142]: Good call.

**Pull request merged**